[DOCS] Reaction Guide (#7495)
Rapptz/discord.py: docs/guide/topics/reactions.rst

.. currentmodule:: discord

.. _guide_topic_reactions

Reactions
==========

Reactions are a feature within Discord where you can add emojis next to messages.
They have many uses, but a common one for bots is for polls or giving feedback to a command without sending a new message.

> Anyone can react to a message.

Reviewer: ```suggestion
Members with the :attr:`~Permissions.add_reactions` can add new reactions to messages. However, this permission is not necessary to add to an existing reaction.
```

The note on line 31 also gets this across, but I feel like it conflicts with this sentence as it is.

Author: i feel like a note would be better than adding that, perhaps just removing the "anyone can react to a message"?

Reviewer: > perhaps just removing the "anyone can react to a message"?

Yeah that seems good.

Adding a reaction
~~~~~~~~~~~~~~~~~~

Adding reactions is quite simple.

To add a custom emoji reaction you can do the following:

.. code-block:: python3

    await message.add_reaction("<:amogus:819046807370989599>")

Which would add a reaction to the message like so:

.. image:: /images/guide/topics/reactions/add_custom_emoji.png
    :scale: 50%

.. note::

    You must have the permission :attr:`~Permissions.read_message_history` to add reactions.
    Additionally, if no one else has reacted to the message using this emoji, you will also need the
    :attr:`~Permissions.add_reactions` permission.

> :meth:`~Message.add_reaction` can also take many types, not just :class:`str`. This includes :class:`~Emoji` or :class:`~PartialEmoji`.

Reviewer: ```suggestion
In addition to :class:`str`, you can also pass :class:`~Emoji`, :class:`~PartialEmoji`, or :class:`~Reaction` to :meth:`~Message.add_reaction`.
```

Exhaustively listing everything that `add_reaction` can accept here might help.

For example, the following shows an alternative way to add a custom emoji reaction:

.. code-block:: python3

    emoji = client.get_emoji(819046807370989599)
    await message.add_reaction(emoji)

> You may notice that if you try to add a standard Discord emoji like ``:smile:``, you'll receive an error.
> This is because to add standard Discord emojis to a message, you must pass their unicode values.
> Because of this, to add 😄 / ``:smile:`` you would do either of the following:

Reviewer: ```suggestion
When adding standard (i.e. not custom) emojis as a reaction, you cannot pass emoji shortcodes,
such as the string ``":smile:"`` in order to react with 😄. The Discord client recognizes these
shortcodes as a typing convenience, but the API is not aware of them. So, in your bot, you must
pass the actual Unicode codepoint as is (e.g., ``"😄"``). This can be done like so:
```

This seems a bit pedantic, but the nomenclature can be ambiguous. "Emoji" can mean actual Unicode codepoints, Discord's custom emoji, or non-standard emoji shortcodes. I tried to reword stuff to make it less confusing here.

Reviewer 2: `:name:` emojis are supported in other contexts as well- such as sending messages.

Reviewer: Unless I'm missing something, shortcodes in messages don't seem to be postprocessed by the API in any way, so the remark "[...] the API is not aware of them" appears to hold:

![image](https://user-images.githubusercontent.com/4206232/158043226-5be8b8a8-ee4c-413a-927d-4b701e4605a6.png)

The client parses them though, of course.

Author: i do not see a reason to introduce client semantics to the guide.

Reviewer: I feel like reducing the amount of potential confusion is desirable here, even at the expense of the result being a bit too wordy or technical. I don't think it's particularly difficult or overtly technical to explain that the API does not recognize shortcodes once it's spelled out, and that emojis are semantically distinct from shortcodes in this way.

I do see that this can be over the top here, though. Oh well pensive

Author: i feel like maybe a "the client and api see emojis in different ways, so you should be doing "X" to make sure you appease the api (in other wording of course).

Reviewer: Oh, that could help, yeah. That's basically what you had in the first place, though, I think. So it's fine without my suggestion, seems.

.. code-block:: python3

>     # pass the unicode emoji via the hexadecimal value
>     await message.add_reaction("\U0001f604")

Reviewer: ```suggestion
    # Pass the emoji codepoint via 32-bit hexadecimal escape.
    # For more information, see: https://docs.python.org/3/howto/unicode.html#the-string-type
    await message.add_reaction("\U0001f604")
```

>     # or
> 
>     # get the unicode emoji via the name (obtained via the unicode database)

Reviewer: ```suggestion
    # Alternatively, use the codepoint's name.
    # See: https://docs.python.org/3/library/unicodedata.html
```

    await message.add_reaction("\N{SMILING FACE WITH OPEN MOUTH AND SMILING EYES}")

This would then result in:

.. image:: /images/guide/topics/reactions/add_standard_emoji.png
    :scale: 50%

You can find the documentation for adding reactions to a message object at :meth:`~Message.add_reaction`.

Removing a reaction
~~~~~~~~~~~~~~~~~~~~~

Removing reactions is similar to adding a reaction, but you must also specify a member to remove the reaction from.
This can be anything that meets the abc :class:`~abc.Snowflake`, like :class:`~discord.Member` or :class:`~discord.Object`.

> This means that you cannot pass simply only an id, and you must instead pass something with an ``id`` attribute.

Reviewer 3: ```suggestion
This means that you cannot simply pass only an id, you must instead pass an object with an ``id`` attribute.
```

IMO sounds a little bit more natural this way.

.. code-block:: python3

    await message.remove_reaction(
        "<:amogus:819046807370989599>",
        guild.me
    )

This removes a custom emoji reaction with the id ``819046807370989599`` added by us (see :attr:`~Guild.me`) from the message:

.. image:: /images/guide/topics/reactions/remove_reaction_before.png
    :scale: 50%

.. image:: /images/guide/topics/reactions/remove_reaction_after.png
    :scale: 50%

> Removing all reactions of a single emoji from a message is also possible: 
> 
> .. code-block:: python3
> 
>     await message.clear_reaction("<:amogus:819046807370989599>")
> 
> It is also possible to remove all reactions from a message, not just a singular member or emoji:

Reviewer 4: This following sentence feels a touch staggered due to the previous sentence (e.g., saying "You can also x. You can also y. You can also z." with no bridges doesn't read very well).

I think it'd probably be best to fix it like:
```suggestion
Removing all reactions of a given emoji from a message is also possible: 

.. code-block:: python3

    await message.clear_reaction("<:amogus:819046807370989599>")

And we can also remove all reactions entirely:
```

.. code-block:: python3

    await message.clear_reactions()

If you have a :class:`~Reaction`, you can also call :meth:`.clear() <Reaction.clear>` on that to clear reactions of that given emoji.

.. note:: 
    
    To remove a reaction that is not your own, you will need the :attr:`~Permissions.manage_messages` permission.

Getting reactions
~~~~~~~~~~~~~~~~~~

The :attr:`~Message.reactions` attribute allows you to get a list of the reactions on the message. This returns a :class:`~Reaction`
per unique emoji on the message.

For example, to get all the users that reacted with a given emoji:

.. code-block:: python3

    # This fetches the first 5 users that reacted with that emoji.
    async for user in reaction.users(limit = 5):
        await channel.send(f"{user} reacted with `{reaction.emoji}")

You can also flatten the users into a list.

.. code-block:: python3

    users = await reaction.users().flatten()
    # users is now a list of User...
    winner = random.choice(users)
    await channel.send(f'{winner} has won the raffle.')

Next Steps
~~~~~~~~~~~

You can find further information about :class:`~Reaction` in its documentation.

> Additionally, :meth:`~Client.wait_for` can pair well with reactions.

Reviewer: Might wanna link to the `wait_for` guide here? It doesn't exist yet though.

Author: i will link it when the guide exists, but currently it does not.

Reviewer 5: The guide now exists and has a reference link, this can be added.
